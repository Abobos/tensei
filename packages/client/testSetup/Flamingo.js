import TextField from '~/fields/Text'
import SelectField from '~/fields/Select'
import HasManyField from '~/fields/HasMany'
import DateTimeField from '~/fields/DateTime'
import TextareaField from '~/fields/Textarea'
import BelongsToField from '~/fields/BelongsTo'

export default {
    getPath: jest.fn(() => 'string'),
    fieldComponents: {
        TextField,
        SelectField,
        HasManyField,
        DateTimeField,
        TextareaField,
        BelongsToField,
        LinkField: TextField,
        NumberField: TextField,
        IntegerField: TextField,
        DateField: DateTimeField,
    },
    library: {
        Notification: { success: jest.fn(), error: jest.fn() },
    },
}
