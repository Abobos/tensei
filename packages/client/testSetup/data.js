export const resources = [
    {
        collection: 'posts',
        defaultPerPage: 1,
        displayInNavigation: true,
        fields: [
            {
                asObjectId: true,
                asString: false,
                attributes: {},
                component: 'TextField',
                creationRules: [],
                databaseField: 'name',
                defaultValue: '',
                description: '',
                fieldName: 'Text',
                inputName: 'name',
                isNullable: true,
                isRelationshipField: false,
                isSearchable: false,
                isSortable: false,
                isUnique: false,
                name: 'Name',
                rules: [],
                showOnCreation: true,
                showOnDetail: true,
                showOnIndex: true,
                showOnUpdate: true,
                sqlDatabaseFieldType: 'string',
                updateRules: [],
            },
            {
                asObjectId: true,
                asString: false,
                attributes: {},
                component: 'TextField',
                creationRules: [],
                databaseField: 'description',
                defaultValue: '',
                description: '',
                fieldName: 'Text',
                inputName: 'description',
                isNullable: true,
                isRelationshipField: false,
                isSearchable: false,
                isSortable: false,
                isUnique: false,
                name: 'description',
                rules: [],
                showOnCreation: true,
                showOnDetail: true,
                showOnIndex: true,
                showOnUpdate: true,
                sqlDatabaseFieldType: 'string',
                updateRules: [],
            },
            {
                asObjectId: true,
                asString: false,
                attributes: {},
                component: 'TextField',
                creationRules: [],
                databaseField: 'content',
                defaultValue: '',
                description: '',
                fieldName: 'Text',
                inputName: 'content',
                isNullable: true,
                isRelationshipField: false,
                isSearchable: false,
                isSortable: false,
                isUnique: false,
                name: 'content',
                rules: [],
                showOnCreation: true,
                showOnDetail: true,
                showOnIndex: true,
                showOnUpdate: true,
                sqlDatabaseFieldType: 'string',
                updateRules: [],
            },
        ],
        group: 'All',
        label: 'Posts',
        slug: 'posts',
        messages: {
            'title.required': 'The title field is required.',
            'publishedAt.required': 'The published at field is required.',
        },
        name: 'Post',
        param: 'posts',
        perPageOptions: [(1, 3, 5)],
        primaryKey: '_id',
    },
]

export const user = {
    email: 'dodo@email.com',
    firstName: 'dozie',
    lastName: 'nwoga',
    password: '$2a$10$d.IeGxbRR4kc1ZxE7u0LSuHMrX9aMlUrbLgLoxqEcVI9I2CyntgV.',
    _id: '5f0d62b4e2fab0431e1d35cf',
}
