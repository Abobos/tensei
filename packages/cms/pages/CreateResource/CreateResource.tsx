import React, { useState, useEffect } from 'react'
import { Heading, Button, AbstractData } from '@tensei/components'

import { Pulse } from '@tensei/components'
import PageWrapper from '../../components/PageWrapper'
import { useParams, Redirect, Link, useHistory } from 'react-router-dom'
import { AxiosError } from 'axios'

interface CreateResourceProps {}

const CreateResource: React.FC<CreateResourceProps> = ({}) => {
    const params = useParams<{
        resource: string
        id: string
    }>()
    const [errors, setErrors] = useState<AbstractData>({})
    const [saving, setSaving] = useState(false)
    const [form, setForm] = useState<AbstractData>({})
    const [booted, setBooted] = useState(false)
    const history = useHistory()

    const resource = window.Tensei.state.resourcesMap[params.resource]

    if (!resource) {
        return <Redirect to={window.Tensei.getPath('404')} />
    }

    const creationFields = resource.fields.filter(field => field.showOnCreation)

    const sidebarFields = creationFields.filter(field => field.sidebar)
    const mainbarFields = creationFields.filter(field => !field.sidebar)

    const sidebarVisible = sidebarFields.length > 0

    const isEditing = !!params.id

    const onErrorCatch = (error: AxiosError) => {
        window.Tensei.error(`Failed saving ${resource.name.toLowerCase()}.`)

        if (error?.response?.status === 422) {
            let errors: AbstractData = {}

            error?.response?.data?.errors.forEach((error: any) => {
                errors[error.field] = error.message
            })

            setErrors(errors)
        }

        setSaving(false)
    }

    const onSave = () => {
        setSaving(true)

        window.Tensei.client[isEditing ? 'patch' : 'post'](
            `/${resource.slug}/${isEditing ? params.id : ''}`,
            form
        )
            .then(() => {
                window.Tensei.success(
                    `${resource.name} ${isEditing ? 'updated' : 'created'}.`
                )

                history.push(
                    window.Tensei.getPath(`resources/${resource.slug}`)
                )
            })
            .catch(onErrorCatch)
    }

    const setDefaultFormData = () => {
        if (!isEditing) {
            let formData: AbstractData = {}

            creationFields.forEach(field => {
                formData[field.inputName] = field.defaultValue
            })

            setForm(formData)
            setBooted(true)

            return
        }

        window.Tensei.client
            .get(`${resource.slug}/${params.id}`)
            .then(({ data }) => {
                let formData: AbstractData = {}

                creationFields.forEach(field => {
                    formData[field.inputName] =
                        data.data[field.inputName] || field.defaultValue
                })

                setForm(formData)
                setBooted(true)
            })
            .catch(() => {
                window.Tensei.error(
                    `${resource.name} with ID ${params.id} not found.`
                )
            })
    }

    useEffect(() => {
        setDefaultFormData()
    }, [])

    const ActionButtons = (
        <div className="flex mt-4 md:mt-0">
            <Link to={window.Tensei.getPath(`resources/${resource.slug}`)}>
                <Button clear>Cancel</Button>
            </Link>
            <Button onClick={onSave} primary loading={saving} className="ml-5">
                Save
            </Button>
        </div>
    )

    return (
        <PageWrapper>
            <header className="flex flex-wrap items-center justify-between">
                <Heading className="w-full md:w-auto" as="h1">
                    {isEditing ? 'Update' : 'Create'} {resource.name}
                </Heading>

                {ActionButtons}
            </header>
            {booted ? (
                <div className="flex flex-wrap md:flex-nowrap mt-6">
                    <div
                        className={`flex flex-col flex-wrap w-full ${
                            sidebarVisible ? 'md:w-4/6' : ''
                        } bg-white border border-tensei-gray-600 rounded-lg p-8 md:mr-8`}
                    >
                        {mainbarFields.map(field => {
                            const Component =
                                window.Tensei.components.form[
                                    field.component.form
                                ] || window.Tensei.components.form.Text

                            return (
                                <div
                                    key={field.inputName}
                                    className={`mb-5 ${
                                        sidebarVisible ? 'w-full' : 'w-3/5'
                                    }`}
                                >
                                    <Component
                                        form={form}
                                        field={field}
                                        resource={resource}
                                        id={field.inputName}
                                        editing={isEditing}
                                        editingId={params.id}
                                        name={field.inputName}
                                        error={errors[field.inputName]}
                                        value={form[field.inputName]}
                                        onChange={(value: any) => {
                                            setForm({
                                                ...form,
                                                [field.inputName]: value
                                            })

                                            setErrors({
                                                ...errors,
                                                [field.inputName]: undefined
                                            })
                                        }}
                                    />
                                </div>
                            )
                        })}
                    </div>
                    {sidebarVisible ? (
                        <div className="w-full md:w-2/6 mt-5 md:mt-0">
                            {sidebarFields.map((field, index) => {
                                const Component =
                                    window.Tensei.components.form[
                                        field.component.form
                                    ] || window.Tensei.components.form.Text

                                return (
                                    <div
                                        key={field.inputName}
                                        className={`bg-white border border-tensei-gray-600 rounded-lg p-8 ${
                                            index === sidebarFields.length - 1
                                                ? ''
                                                : 'mb-8'
                                        }`}
                                    >
                                        <Component
                                            form={form}
                                            field={field}
                                            editing={isEditing}
                                            resource={resource}
                                            editingId={params.id}
                                            id={field.inputName}
                                            name={field.inputName}
                                            error={errors[field.inputName]}
                                            value={form[field.inputName]}
                                            onChange={(value: any) => {
                                                setForm({
                                                    ...form,
                                                    [field.inputName]: value
                                                })

                                                setErrors({
                                                    ...errors,
                                                    [field.inputName]: undefined
                                                })
                                            }}
                                        />
                                    </div>
                                )
                            })}
                        </div>
                    ) : null}
                </div>
            ) : (
                <div className="w-full flex justify-center py-12">
                    <Pulse dotClassName="bg-tensei-primary" />
                </div>
            )}

            <div className="md:hidden">{ActionButtons}</div>
        </PageWrapper>
    )
}

export default CreateResource
