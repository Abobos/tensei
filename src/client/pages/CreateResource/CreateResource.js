import React from 'react'
import { withResources } from 'store/resources'
import { Text } from 'office-ui-fabric-react/lib/Text'
import { DefaultButton, PrimaryButton } from 'office-ui-fabric-react/lib/Button'

class CreateResource extends React.Component {
    state = this.defaultState()

    defaultState() {
        return {
            form: {},
            formInitialized: false,
            resource: this.findResource(),
            editingState: !!this.props.match.params.resourceId,
        }
    }

    findResource() {
        return this.props.resources.find(
            (resource) => resource.param === this.props.match.params.resource
        )
    }

    componentDidMount() {
        this.initializeForm()
    }

    initializeForm = () => {
        this.setState({
            formInitialized: true,
            form: this.getDefaultFormState(),
        })
    }

    getUpdateFields = () =>
        this.state.resource.fields.filter((field) => field.showOnUpdate)

    getCreationFields = () =>
        this.state.resource.fields.filter((field) => field.showOnCreation)

    getResourceFields = () =>
        this.state.editingState
            ? this.getUpdateFields()
            : this.getCreationFields()

    getDefaultFormState = () =>
        this.state.editingState
            ? this.getDefaultEditingFormState()
            : this.getDefaultCreationFormState()

    getDefaultEditingFormState = () => {}

    resetForm = () => this.setState({
        form: this.getDefaultFormState()
    })

    getDefaultCreationFormState = () => {
        const form = {}

        this.getCreationFields().forEach((field) => {
            form[field.inputName] = field.defaultValue
        })

        return form
    }

    renderResourceField = (resourceField) => {
        const Component = Flamingo.fieldComponents[resourceField.component]

        return (
            <div key={resourceField.inputName} className="mb-3">
                <Component
                    field={resourceField}
                    onChange={console.log}
                    label={resourceField.name}
                    value={this.state.form[resourceField.inputName]}
                />
            </div>
        )
    }

    render() {
        const { resource } = this.state

        return (
            <React.Fragment>
                <header className="flex flex-wrap items-center justify-between">
                    <Text variant="xLarge">Create {resource.name}</Text>

                    <div className="w-full md:w-auto mt-4 md:mt-0">
                        <DefaultButton onClick={this.resetForm} className="mr-3">Reset</DefaultButton>

                        <PrimaryButton>
                            <Text>Create {resource.name}</Text>
                        </PrimaryButton>
                    </div>
                </header>

                <div className="w-full flex flex-wrap mt-10">
                    <div className="w-full md:w-1/4 flex flex-col mb-5 md:mb-0">
                        <Text variant="large">{resource.name}</Text>
                        <Text variant="medium" className="opacity-75">
                            Put in information about the new{' '}
                            {resource.name.toLowerCase()}
                        </Text>
                    </div>

                    <div className="w-full md:w-2/4 bg-white shadow px-6 py-6">
                        {this.getResourceFields().map(this.renderResourceField)}
                    </div>
                </div>
            </React.Fragment>
        )
    }
}

export default withResources(CreateResource)
