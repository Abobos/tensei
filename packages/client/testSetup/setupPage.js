import React from 'react'
import { render } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import { MemoryRouter } from 'react-router-dom'

import Auth from '~/store/auth'
import Resources from '~/store/resources'
import { resources, user } from './data'

const history = createMemoryHistory()

const setupPage = (
    initialEntries,
    initialIndex,
    Component,
    withAuth = true
) => {
    const props = { resources }
    return render(
        <MemoryRouter
            initialIndex={initialIndex}
            initialEntries={initialEntries}
        >
            <Auth.Provider value={withAuth ? [user, () => jest.fn()] : []}>
                <Resources.Provider
                    value={{
                        resources: resources,
                    }}
                >
                    <Component {...props} location={history.location} />
                </Resources.Provider>
            </Auth.Provider>
        </MemoryRouter>
    )
}
export default setupPage
