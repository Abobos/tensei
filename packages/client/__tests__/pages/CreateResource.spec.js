import React from 'react'
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import setupPage from '~/testSetup/setupPage'
import Auth from '~/store/auth'
import { resources } from '~/testSetup/data'
import FlamingoMock from '~/testSetup/Flamingo'
import CreateResource from '~/pages/CreateResource'

const WithAuthComponent = (props) => (
    <Auth.Provider
        value={{
            authorizedToCreate: jest.fn(() => Promise.resolve(true)),
        }}
    >
        <CreateResource {...props} />
    </Auth.Provider>
)

describe('Test the create resource page', () => {
    beforeEach(() => {
        window.Flamingo = FlamingoMock
    })

    test('create resource page should match snapshot', () => {
        const match = { params: { resource: 'posts' } }
        const props = {
            match,
            resources,
        }

        const { asFragment } = render(<WithAuthComponent {...props} />)

        expect(asFragment()).toMatchSnapshot()
    })

    test('can create a resource i.e post', () => {
        const match = { params: { resource: 'posts' } }
        const props = {
            match,
            resources,
            history: { push: jest.fn() },
        }

        window.Flamingo = {
            ...FlamingoMock,
            request: { post: jest.fn(() => Promise.resolve(true)) },
        }
        render(<WithAuthComponent {...props} />)

        expect(screen.getByTestId('resource-title')).toBeInTheDocument()
        expect(screen.getAllByText(/Create post/i)).toHaveLength(2)

        const nameField = screen.getByLabelText('name')
        const descriptionField = screen.getByLabelText('description')
        const contentField = screen.getByLabelText('content')
        const submitBtn = screen.getByTestId('submit-button')

        fireEvent.change(nameField, { target: { value: 'new post' } })
        fireEvent.change(descriptionField, {
            target: { value: 'new description' },
        })
        fireEvent.change(contentField, { target: { value: 'fresh content' } })

        fireEvent.click(submitBtn)

        expect(window.Flamingo.request.post).toHaveBeenCalled()
    })
})
