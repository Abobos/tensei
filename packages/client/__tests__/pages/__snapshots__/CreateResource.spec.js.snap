// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the create resource page create resource page should match snapshot 1`] = `
<DocumentFragment>
  <header
    class="flex flex-wrap items-center justify-between"
  >
    <h1
      class="Heading__Heading___83D3K mb-0"
      data-test-id="cf-ui-heading"
      data-testid="resource-title"
    >
      Create post
    </h1>
    <div
      class=""
    >
      <button
        class="Button__Button___1ZfFj a11y__focus-border--default___60AXp mr-3 Button__Button--muted___2Wair"
        data-test-id="cf-ui-button"
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <span
            class="Button__Button__label___3tcOj"
          >
            Reset
          </span>
        </span>
      </button>
      <button
        class="Button__Button___1ZfFj a11y__focus-border--default___60AXp Button__Button--primary___JImeO"
        data-test-id="cf-ui-button"
        data-testid="submit-button"
        type="button"
      >
        <span
          class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
          tabindex="-1"
        >
          <span
            class="Button__Button__label___3tcOj"
          >
            Create post
          </span>
        </span>
      </button>
    </div>
  </header>
  <div
    class="w-full flex flex-wrap mt-10"
  >
    <div
      class="w-full md:w-1/3 flex flex-col mb-5 md:mb-0"
    >
      <h2
        class="Subheading__Subheading___2mA9j"
        data-test-id="cf-ui-subheading"
        variant="large"
      >
        Post
      </h2>
      <p
        class="Paragraph__Paragraph___2aO-9"
        data-test-id="cf-ui-paragraph"
      >
        Put in information about the new post
      </p>
    </div>
    <div
      class="w-full md:w-2/3"
    >
      <div
        class="w-full md:w-5/6 bg-gray-lightest p-6 mb-3"
      >
        <div
          class="mb-3"
        >
          <div
            class="TextField__TextField___Sf6eo TextField__TextField--full___11DBK"
            data-test-id="cf-ui-text-field"
            description=""
          >
            <div
              class="TextField__TextField__label-wrapper___2-MJT"
            >
              <label
                class="FormLabel__FormLabel___3d6zQ"
                data-test-id="cf-ui-form-label"
                for="name"
              >
                Name
              </label>
            </div>
            <div
              class="TextInput__TextInput___36-K- TextInput__TextInput--full___1EJEW"
            >
              <input
                aria-label="name"
                class="TextInput__TextInput__input___27vDB a11y__focus-border--default___60AXp"
                data-test-id="cf-ui-text-input"
                id="name"
                name="name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="mb-3"
        >
          <div
            class="TextField__TextField___Sf6eo TextField__TextField--full___11DBK"
            data-test-id="cf-ui-text-field"
            description=""
          >
            <div
              class="TextField__TextField__label-wrapper___2-MJT"
            >
              <label
                class="FormLabel__FormLabel___3d6zQ"
                data-test-id="cf-ui-form-label"
                for="description"
              >
                description
              </label>
            </div>
            <div
              class="TextInput__TextInput___36-K- TextInput__TextInput--full___1EJEW"
            >
              <input
                aria-label="description"
                class="TextInput__TextInput__input___27vDB a11y__focus-border--default___60AXp"
                data-test-id="cf-ui-text-input"
                id="description"
                name="description"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="mb-3"
        >
          <div
            class="TextField__TextField___Sf6eo TextField__TextField--full___11DBK"
            data-test-id="cf-ui-text-field"
            description=""
          >
            <div
              class="TextField__TextField__label-wrapper___2-MJT"
            >
              <label
                class="FormLabel__FormLabel___3d6zQ"
                data-test-id="cf-ui-form-label"
                for="content"
              >
                content
              </label>
            </div>
            <div
              class="TextInput__TextInput___36-K- TextInput__TextInput--full___1EJEW"
            >
              <input
                aria-label="content"
                class="TextInput__TextInput__input___27vDB a11y__focus-border--default___60AXp"
                data-test-id="cf-ui-text-input"
                id="content"
                name="content"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="mb-3"
        >
          <div
            class="TextField__TextField___Sf6eo TextField__TextField--full___11DBK"
            data-test-id="cf-ui-text-field"
            description=""
          >
            <div
              class="TextField__TextField__label-wrapper___2-MJT"
            >
              <label
                class="FormLabel__FormLabel___3d6zQ"
                data-test-id="cf-ui-form-label"
                for="visuals"
              >
                Visuals
              </label>
            </div>
            <div
              class="TextInput__TextInput___36-K- TextInput__TextInput--full___1EJEW"
            >
              <input
                aria-label="visuals"
                class="TextInput__TextInput__input___27vDB a11y__focus-border--default___60AXp"
                data-test-id="cf-ui-text-input"
                id="visuals"
                name="visuals"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
